modules/RsIskraMeter/src/utils.rs

```diff
@@ -0,0 +1,166 @@
+use anyhow::Result;
+use rand::Rng;
+
+pub fn to_8_string(input: &[u16]) -> Result<String> {
+    let u8_slice = input.iter().flat_map(|&x| u16::to_be_bytes(x)).collect();
+    Ok(String::from_utf8(u8_slice)?.trim().to_string())
+}
+
+pub fn counter(regs: [u16; 2], exp: u16) -> f64 {
+    let measurement = (regs[0] as i32) << 16 | ((regs[1] as i32) & 0xFFFF);
+    measurement as f64 * 1.0e1_f64.powi(exp as i32)
+}
+
+/// Unsigned Measurement (32 bit)
+/// Decade Exponent (Signed 8 bit)
+/// Binary Signed value (24 bit)
+pub fn from_t5_format(regs: [u16; 2]) -> f64 {
+    let exp = ((regs[0] >> 8) & 0xFF) as i8;
+    let value = (regs[0] & 0xFF) as u8;
+    let measurement = ((value as u32) << 16) | ((regs[1] as u32) & 0xFFFF);
+    measurement as f64 * 1.0e1_f64.powi(exp as i32)
+}
+
+/// Signed Measurement (32 bit)
+/// Decade Exponent (Signed 8 bit)
+/// Binary Signed value (24 bit)
+pub fn from_t6_format(regs: [u16; 2]) -> f64 {
+    let exp = ((regs[0] >> 8) & 0xFF) as i8;
+    let value = (regs[0] & 0xFF) as i8;
+    let measurement = ((value as i32) << 16) | ((regs[1] as i32) & 0xFFFF);
+    measurement as f64 * 1.0e1_f64.powi(exp as i32)
+}
+
+pub fn string_to_vec(input: &str) -> Vec<u16> {
+    input
+        .as_bytes()
+        .chunks(2)
+        .map(|chunk| {
+            let mut value = (chunk[0] as u16) << 8;
+            if chunk.len() > 1 {
+                value |= chunk[1] as u16 & 0xFF;
+            }
+            value
+        })
+        .collect()
+}
+
+pub fn create_random_meter_session_id() -> String {
+    let start = format!("{:06X}", rand::thread_rng().gen_range(0..=0xFFFFFF));
+    let hex_time = format!(
+        "{:X}",
+        std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+    );
+    let end = format!("{:06X}", rand::thread_rng().gen_range(0..=0xFFFFFF));
+
+    start + &hex_time + &end
+}
+
+pub fn to_hex_string(input: Vec<u16>) -> String {
+    input
+        .into_iter()
+        .flat_map(|value| value.to_be_bytes())
+        .map(|value| format!("{value:02X}"))
+        .collect::<Vec<_>>()
+        .concat()
+}
+
+pub fn create_ocmf(signed_meter_values: String, signature: String) -> String {
+    format!("OCMF|{}|{{\"SD\":\"{}\"}}", signed_meter_values, signature)
```

Comment:
A general comment because at this place everything seems to be put together:

`The sections within the data format are separated with the special pipe character "|". Within the sections, the format corresponds to the JSON notation. Thus, the use of the special pipe character "|" is not allowed within the sections.`

From my understanding we need to filter all the "|" from external input strings.

Reply:
I don't think its like that - as long the `|` is in a escaped string within the JSON it should be fine

Reply:
I've checked some sample data containing `|` and the ocmf transparentzsoftware tool and indeed it will not accept any data with `|` ...

However, here we can't remove the `|` since the signature will not fit anymore. To my understanding the only source for a `|` would be from user's config (e.x. when specifying `GV` or similar) - so I would add a check there and filter the `|` there